src/Data/Formatter/DateTime.purs

```diff
@@ -0,0 +1,426 @@
+  map f (MonthFull a) = MonthFull $ f a
+  map f (MonthShort a) = MonthShort $ f a
+  map f (MonthTwoDigits a) = MonthTwoDigits $ f a
+  map f (DayOfMonth a) = DayOfMonth $ f a
+  map f (UnixTimestamp a) = UnixTimestamp $ f a
+  map f (DayOfWeek a) = DayOfWeek $ f a
+  map f (Hours24 a) = Hours24 $ f a
+  map f (Hours12 a) = Hours12 $ f a
+  map f (Meridiem a) = Meridiem $ f a
+  map f (Minutes a) = Minutes $ f a
+  map f (Seconds a) = Seconds $ f a
+  map f (Milliseconds a) = Milliseconds $ f a
+  map f (Placeholder str a) = Placeholder str $ f a
+  map f End = End
+
+type Formatter = Mu FormatterF
+
+printFormatterF
+  ∷ ∀ a
+  . (a → String)
+  → FormatterF a
+  → String
+printFormatterF cb = case _ of
+  YearFull a → "YYYY" <> cb a
+  YearTwoDigits a → "YY" <> cb a
+  YearAbsolute a → "Y" <> cb a
+  MonthFull a → "MMMM" <> cb a
+  MonthShort a → "MMM" <> cb a
+  MonthTwoDigits a → "MM" <> cb a
+  DayOfMonth a → "DD" <> cb a
+  UnixTimestamp a → "X" <> cb a
+  DayOfWeek a → "E" <> cb a
+  Hours24 a → "HH" <> cb a
+  Hours12 a → "hh" <> cb a
+  Meridiem a → "a" <> cb a
+  Minutes a → "mm" <> cb a
+  Seconds a → "ss" <> cb a
+  Milliseconds a → "SSS" <> cb a
+  Placeholder s a → s <> cb a
+  End → ""
+
+printFormatter ∷ Formatter → String
+printFormatter f = printFormatterF printFormatter $ unroll f
+
+parseFormatString ∷ String → Either String Formatter
+parseFormatString s =
+  lmap (\(P.ParseError {message}) → message) $ P.runParser s formatParser
+
+
+placeholderContent ∷ P.Parser String String
+placeholderContent =
+  map Str.fromCharArray
+    $ PC.try
+    $ Arr.some
+    $ PS.noneOf
+    $ Str.toCharArray "YQMDXWEHhamsS"
+
+formatterFParser
+  ∷ ∀ a
+  . P.Parser String a
+  → P.Parser String (FormatterF a)
+formatterFParser cb =
+  PC.choice
+    [ (PC.try $ PS.string "YYYY") *> map YearFull cb
+    , (PC.try $ PS.string "YY") *> map YearTwoDigits cb
+    , (PC.try $ PS.string "Y") *> map YearAbsolute cb
+    , (PC.try $ PS.string "MMMM") *> map MonthFull cb
+    , (PC.try $ PS.string "MMM") *> map MonthShort cb
+    , (PC.try $ PS.string "MM") *> map MonthTwoDigits cb
+    , (PC.try $ PS.string "DD") *> map DayOfMonth cb
+    , (PC.try $ PS.string "E") *> map DayOfWeek cb
+    , (PC.try $ PS.string "HH") *> map Hours24 cb
+    , (PC.try $ PS.string "hh") *> map Hours12 cb
+    , (PC.try $ PS.string "a") *> map Meridiem cb
+    , (PC.try $ PS.string "mm") *> map Minutes cb
+    , (PC.try $ PS.string "ss") *> map Seconds cb
+    , (PC.try $ PS.string "SSS") *> map Milliseconds cb
+    , (Placeholder <$> placeholderContent <*> cb)
+    , (PS.eof $> End)
+    ]
+
+formatParser ∷ P.Parser String Formatter
+formatParser =
+  Lazy.fix \f → map roll $ formatterFParser f
+
+formatF
+  ∷ ∀ a
+  . (a → String)
+  → DT.DateTime
+  → FormatterF a
+  → String
+formatF cb dt@(DT.DateTime d t) = case _ of
+  YearFull a →
+    (show $ fromEnum $ D.year d) <> cb a
+  YearTwoDigits a →
+    show ((fromEnum $ D.year d) - 1900) <> cb a

```

Comment:
Ok, I'll switch to something like `\x -> if ix > 30 then 1900 + x else 2000 + x`